OpenUrlRedirect: Expand safe URL flow configuration

sauyon opened the pull request:
Tested this on LGTM.com and it fixes all but two of the obvious false positives.

The last two are [these](https://lgtm.com/projects/g/caddyserver/caddy/snapshot/89b6c8ef3926b30ed1b8dca64f8ebd3301c9321f/files/caddyhttp/staticfiles/fileserver.go?sort=name&dir=ASC&mode=heatmap#xf26e02affcbaab40:1) [results](https://lgtm.com/projects/g/caddyserver/caddy/snapshot/89b6c8ef3926b30ed1b8dca64f8ebd3301c9321f/files/caddyhttp/staticfiles/fileserver.go?sort=name&dir=ASC&mode=heatmap#xc9a7505ef87e05cd:1) on `caddyserver/caddy`, for which the source, which I'd expect to be on line 102 of `fileserver.go`, is strangely not found. I'll probably do some digging into this eventually, but I thought I'd leave it here for now since it seems like this may be a bit more involved.

max-schaefer commented on ql/src/semmle/go/security/OpenUrlRedirect.qll:
```diff
@@ -54,16 +54,41 @@ module OpenUrlRedirect {
+        (frn.getFieldName() = "Host" or frn.getFieldName() = "Path")
+      )
+      or
+      exists(Write w, Field f, SsaWithFields v | f.hasQualifiedName("net/url", "URL", "Path") |
```
Could you add a (code) comment explaining this step?

sauyon replied:
Hm, these should really have been on `Host`; I'll change that (and add a comment).

max-schaefer commented on ql/src/semmle/go/security/OpenUrlRedirect.qll:
```diff
@@ -54,16 +54,41 @@ module OpenUrlRedirect {
+    }
+
+    override predicate isBarrierOut(DataFlow::Node node) {
+      exists(Write w, Field f | f.hasQualifiedName("net/url", "URL", "Path") |
```
Again, a comment would be helpful.

sauyon commented on ql/src/semmle/go/security/OpenUrlRedirect.qll:
```diff
@@ -54,16 +54,41 @@ module OpenUrlRedirect {
       or
+      TaintTracking::functionModelStep(any(Path::PathManipulatingFunction pmf), pred, succ)
+      or
+      TaintTracking::functionModelStep(any(StringRightTrimmer rt), pred, succ)
```
My thinking was that a lot of the taint functions, like concatenation or prefix removal, might make the URL unsafe, and in general I didn't expect that to be a very common use case.

I think here and below the important things to blacklist are the `fieldRead` steps and `UrlGetter` steps, so I'll do that and run a dist-compare to make sure we don't lose CVEs.

max-schaefer replied:
> My thinking was that a lot of the taint functions, like concatenation or prefix removal, might make the URL unsafe

That's a fair point. A whitelist makes sense in that case.

max-schaefer replied:
Thinking about this some more, perhaps a combined black-and-white list would make sense? We could say that we are willing to accept any taint step except `fieldRead` steps and function-model steps, but for the latter specifically allow the above cases.

sauyon replied:
I thought about this too and I think usually if the developer is using the entire URL object it's unlikely to be in an unsafe redirect, just because it would have its host set explicitly.

I think that therefore making a blacklist, just for the parts of the URL which are explicitly user-controlled (like the path / query), makes sense.

max-schaefer replied:
OK, sounds good.

max-schaefer replied:
Can you expand on that a little?

sauyon replied:
I was just wondering whether you knew how much browsers normalize URLs before sending them (and I suppose whether something like `https://myhost.com//malicious.com` could be sent, resulting in the URL path being `//malicious.com`).

max-schaefer replied:
Oh, right, I believe that will be normalised to `/malicious.com`. You can send weird paths with curl, but that's out of scope for open-redirect, where the attack vector would go through an unsuspecting user clicking on a benign-looking link in their browser.

sauyon replied:
Makes sense, thanks.

max-schaefer commented:
> I'll do that and run a dist-compare

Just to clarify: we are still waiting for an evaluation, yes?

sauyon commented:
> Just to clarify: we are still waiting for an evaluation, yes?

This must have failed, I'll check on it.

sauyon commented:
> ...You can send weird paths with curl, but that's out of scope for open-redirect, where the attack vector would go through an unsuspecting user clicking on a benign-looking link in their browser.

Does that mean that Location and Referer headers can both be considered safe in this context as well? I noticed that a few of the results remaining use those values.

max-schaefer commented:
Probably; see my explanation of the typical attack scenario above.

sauyon commented:
Ok, I'll remove those as sources.

sauyon commented:
I've added a few extra things, including something I'd forgotten to change from the LGTM version of the query.

I've decided to remove all sources from request headers, because it seems to me that those wouldn't be as attacker-controllable. I've also changed the standard library models for `Header.Get` and `Header.Values` to allow taint through instead of flagging them as unsafe by default.

An [evaluation]() shows that we now also fix a false positive in our default suite :tada:

max-schaefer commented on ql/src/semmle/go/security/OpenUrlRedirect.qll:
```diff
@@ -41,9 +41,21 @@ module OpenUrlRedirect {
       TaintTracking::localTaintStep(pred, succ) and not TaintTracking::fieldReadStep(pred, succ)
+      or
+      // propagate to a URL when its host is assigned to
+      exists(Write w, Field f, SsaWithFields v | f.hasQualifiedName("net/url", "URL", "Host") |
```
It also looks to me like this disjunct makes the first one above ("A write to URL.Host") obsolete; or am I missing something?

sauyon replied:
Indeed.

sauyon commented:
Ok, added a change note.

max-schaefer commented on change-notes/1.24/analysis-go.md:
```diff
@@ -21,11 +21,12 @@ The CodeQL library for Go now contains a folder of simple "cookbook" queries tha
+| Reflected cross-site scripting (`go/reflected-xss`)                                                                                 | Fewer results         | Untrusted input flowing into an HTTP header definition or into an `fmt.Fprintf` call with a constant prefix is no longer flagged, since it is in both cases often harmless. |
+| Useless assignment to field (`go/useless-assignment-to-field`)                                                                      | Fewer false positives | The query now conservatively handles fields promoted through embedded pointer types.                                                                                        |
+| Bitwise exclusive-or used like exponentiation (`go/mistyped-exponentiation`)                                                        | Fewer false positives | The query now identifies when the value of an xor is assigned to a mask object, and excludes such results.                                                                  |
+| Open URL redirection based on unvalidated user input may cause redirection to malicious websiets (`go/unvalidated-url-redirection`) | Fewer false positives | The query now identifies some sources that are not attacker-controlled, and excludes results with such sources.                                                             |
```
```suggestion
| Open URL redirect (`go/unvalidated-url-redirection`) | Fewer false positives | The query now identifies some sources that are not attacker-controlled, and excludes results with such sources.                                                             |
```

sauyon replied:
Ah, right.
